Fix AttributeError for pickle load of Figure class

wsykala opened the pull request:
## PR Summary

When saving a figure with the `pickle` module in one `python` process and then loading it in a different process a following error appears:
```
/figure.py", line 2911, in __setstate__
    mgr = plt._backend_mod.new_figure_manager_given_figure(num, self)
AttributeError: 'NoneType' object has no attribute 'new_figure_manager_given_figure'
```

This PR replaces the `plt._backend_mod` with the call to `plt._get_backend_mod()` in the `Figure.__setstate__` method. This change ensures that the `_backend_mod` is not `None` during `pickle` load.

A reproducible example is provided in the comments

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

wsykala commented:
Reproducible example:

```
# save.py
import numpy as np
import matplotlib.pyplot as plt
import pickle

fig_handle = plt.figure()
x = np.linspace(0,2*np.pi)
y = np.sin(x)

# Save figure handle to disk
with open('sinus.pickle','wb') as file
    pickle.dump(fig_handle, file)
```

```
# read.py
import matplotlib.pyplot as plt
import pickle
import numpy as np

# Load figure from disk and display
with open('sinus.pickle', 'rb') as blob:
    fig = pickle.load(blob)
```

Outcome: 
```
$ python save.py
$ python read.py
Traceback (most recent call last):
  File "/Users/wsykala/projects/playground/read.py", line 6, in <module>
    fig_handle = pl.load(open('sinus.pickle','rb'))
  File "/Users/wsykala/miniconda3/envs/matplotlib/lib/python3.9/site-packages/matplotlib/figure.py", line 2911, in __setstate__
    mgr = plt._backend_mod.new_figure_manager_given_figure(num, self)
AttributeError: 'NoneType' object has no attribute 'new_figure_manager_given_figure'
```

QuLogic commented:
We already have tests for pickling; I assume they didn't catch this because it's all in the same process.

We definitely need a test to be sure that this is working.

tacaswell commented:
This is another fallout from making backend setup lazier in pyplot (previously we would call `switch_backend` on import which prevented this).  

I suspect https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html#the-tmp-path-fixture +https://github.com/matplotlib/matplotlib/blob/4455dc97fdd993e169485190f5f0d9949d1d7ddd/lib/matplotlib/testing/__init__.py#L53-L66 is what we need to test this.

wsykala commented:
@tacaswell @QuLogic 
Added a test for this behaviour and extracted the code that was duplicated. The test scenario is as follows: 

We create and pickle the figure in the main process. Next we run the `pickle.load` and the dump it to stdout in the subprocess.  After that the binary representation of the figure is loaded from stdout using `ast`, and we compare it the same way as in `test_complete`.

I am not sure why the `circleci` job for building docs failed though.

Also one more point. When creating tests I came across a bug. I described it in this issue #23471

jklymak commented:
The CI errors are 

```
/home/circleci/project/lib/matplotlib/axes/_base.py:docstring of matplotlib.axes._base._AxesBase.set_prop_cycle:14: WARNING: py:obj reference target not found: cycler.Cycler
/home/circleci/project/lib/matplotlib/axes/_base.py:docstring of matplotlib.axes._base._AxesBase.set_prop_cycle:16: WARNING: py:obj reference target not found: cycler.Cycler
/home/circleci/project/lib/matplotlib/axes/_base.py:docstring of matplotlib.axes._base._AxesBase.set_prop_cycle:23: WARNING: py:obj reference target not found: cycler.Cycler
/home/circleci/project/lib/matplotlib/axes/_base.py:docstring of matplotlib.axes._base._AxesBase.set_prop_cycle:23: WARNING: py:obj reference target not found: cycler.cycler
/home/circleci/project/lib/matplotlib/axes/_base.py:docstring of matplotlib.axes._base._AxesBase.set_prop_cycle:55: WARNING: py:obj reference target not found: cycler.cycler
/home/circleci/project/lib/matplotlib/rcsetup.py:docstring of matplotlib.rcsetup.cycler:2: WARNING: py:obj reference target not found: cycler.Cycler
/home/circleci/project/lib/matplotlib/rcsetup.py:docstring of matplotlib.rcsetup.cycler:2: WARNING: py:func reference target not found: cycler.cycler
/home/circleci/project/lib/matplotlib/rcsetup.py:docstring of matplotlib.rcsetup.cycler:11: WARNING: py:obj reference target not found: cycler.Cycler
/home/circleci/project/lib/matplotlib/rcsetup.py:docstring of matplotlib.rcsetup.cycler:13: WARNING: py:obj reference target not found: cycler.Cycler
/home/circleci/project/lib/matplotlib/rcsetup.py:docstring of matplotlib.rcsetup.cycler:17: WARNING: py:obj reference target not found: cycler.Cycler
/home/circleci/project/lib/matplotlib/rcsetup.py:docstring of matplotlib.rcsetup.cycler:38: WARNING: py:class reference target not found: cycler.Cycler
```

But I don't see that this PR should have caused that....

tacaswell commented:
I think the docs failure is related to hosting issues (which were resolved by our providers), re-restarted docs build.

QuLogic commented:
Thanks @wsykala! Congratulations on your first PR to Matplotlib :tada: We hope to hear from you again.

tacaswell commented:
There are same pretty big changes in the test file on main that are not an 3.5.x (for example the pickle test went to an figures equal test from an image comparison test!).

I'm inclined to backport this without backporting the tests.  It is not ideal, but I think it is less bad than re-writing the tests to not confilct (and dealing with a messy merge up) or implicitly backporting a bunch of other changes to the tests.

tacaswell commented:
Discussed on the call.  Plan: copy the tests back and accept the code duplication and clean it up on merge up.
